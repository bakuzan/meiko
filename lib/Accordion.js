import classNames from 'classnames';
import React, { useState, useEffect } from 'react';

import { usePrevious } from './hooks/usePrevious';
import styles from './styles/Accordion';

function AccordionToggle({ className, heading, onToggle }) {
  return (
    <button type="button" className={className} onClick={onToggle}>
      {heading}
    </button>
  );
}

function Accordion({
  children,
  className,
  contentProps = {},
  defaultIsCollapsed,
  headingProps = {},
  heading,
  toggleComponent = AccordionToggle,
  onToggle,
  ...pass
}) {
  const [isCollapsed, setCollapsed] = useState(defaultIsCollapsed ?? true);
  const prevCollapsed = usePrevious(isCollapsed);

  const Toggler = toggleComponent;

  useEffect(() => {
    const hasPrev = prevCollapsed !== undefined;
    const hasNew = prevCollapsed !== isCollapsed;

    if (onToggle && hasPrev && hasNew) {
      onToggle(isCollapsed);
    }
  }, [onToggle, isCollapsed, prevCollapsed]);

  return (
    <div
      {...pass}
      className={classNames('accordion', styles.accordion, className)}
    >
      <div
        {...headingProps}
        className={classNames(
          'accordion__heading',
          styles.accordion__heading,
          headingProps.className
        )}
      >
        <Toggler
          className={classNames(
            'accordion__toggle',
            styles.accordion__toggle,
            isCollapsed && [
              'accordion__toggle--checked',
              styles.accordion__toggle_collapsed
            ]
          )}
          heading={heading}
          onToggle={() => setCollapsed((p) => !p)}
        />
      </div>
      <div
        {...contentProps}
        aria-hidden={isCollapsed}
        className={classNames(
          'accordion__content',
          styles.accordion__content,
          contentProps.className,
          isCollapsed && [
            'accordion__content--collapsed',
            styles.accordion__content_collapsed
          ]
        )}
      >
        {typeof children !== 'function' ? children : children(isCollapsed)}
      </div>
    </div>
  );
}

Accordion.propTypes = {
  heading: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.node
  ]).isRequired,
  defaultIsCollapsed: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.node
  ]),
  onToggle: PropTypes.func
};

export default Accordion;
