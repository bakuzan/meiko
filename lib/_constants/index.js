export { default as Strings } from './strings';
export { default as Types } from './types';
export { default as Urls } from './urls';
export { default as Icons } from './icons';
export { PositionEnum, KeyCodes, OPEN_KEYS, CLOSE_KEYS } from './enums';
