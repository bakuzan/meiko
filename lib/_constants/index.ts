import _Strings from './strings';
import _Types from './types';
import _Urls from './urls';
import _Icons from './icons';
import * as _Enums from './enums';

export const Strings = _Strings;
export const Types = _Types;
export const Urls = _Urls;
export const Icons = _Icons;
export const Enums = _Enums;
