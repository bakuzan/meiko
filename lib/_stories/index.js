import React from 'react';
// import { storiesOf } from '@storybook/react';

import './styles/index.scss';

// storiesOf('Index', module).add('basic', () => <div>Exists to load styles</div>);

export const storyStyle = {
  maxWidth: '300px',
  margin: 'auto',
  marginTop: '20px'
};

const styles = { display: 'flex', justifyContent: 'center', marginTop: '30px' };
export const CenterDecorator = (storyFn) => (
  <div style={styles}>{storyFn()}</div>
);
