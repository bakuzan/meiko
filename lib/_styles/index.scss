@import './partials/variables.scss';
@import './box-model.scss';
@import './list.scss';
@import './input-controls.scss';

.button-group {
  display: flex;
  justify-content: space-around;
  align-items: center;
  padding: 5px;
  margin: 5px 0;

  &.centered {
    justify-content: center;
  }
  &.right-aligned {
    justify-content: flex-end;
  }
}
