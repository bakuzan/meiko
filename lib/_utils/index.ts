import adjustDateMonth from 'ayaka/adjustDateMonth';
import adjustDateYear from 'ayaka/adjustDateYear';
import { capitalise } from 'ayaka/capitalise';
import checkIfDatePartsMatch from 'ayaka/checkIfDatePartsMatch';
import convertToBase64 from 'ayaka/convertToBase64';
import createListeners from 'ayaka/createListeners';
import debounce from 'ayaka/debounce';
import formatDateForInput from 'ayaka/formatDateForInput';
import generateUniqueId from 'ayaka/generateUniqueId';
import getDaysInMonthForDate from 'ayaka/getDaysInMonthForDate';
import getDifferenceFromMonday from 'ayaka/daysDifferenceFromMonday';
import getDifferenceFromSunday from 'ayaka/daysDifferenceFromSunday';
import getElementCoordinates from 'ayaka/getElementCoordinates';
import getEventValue from 'ayaka/getEventValue';
import getFirstDateOfMonth from 'ayaka/getFirstDateOfMonth';
import getLastDateOfMonth from 'ayaka/getLastDateOfMonth';
import getMonthName from 'ayaka/getMonthName';
import { getTimeoutSeconds } from 'ayaka/getTimeout';
import isArray from 'ayaka/isArray';
import { isBefore, isBeforeOrEqual } from 'ayaka/isBefore';
import isNumber from 'ayaka/isNumber';
import isObject from 'ayaka/isObject';
import isString from 'ayaka/isString';
import { lighten, darken } from 'ayaka/palette';

export {
  adjustDateMonth,
  adjustDateYear,
  capitalise,
  convertToBase64,
  createListeners,
  debounce,
  generateUniqueId,
  getElementCoordinates,
  getEventValue,
  getTimeoutSeconds,
  isArray,
  isNumber,
  isObject,
  isString,
  checkIfDatePartsMatch,
  formatDateForInput,
  getMonthName,
  getDaysInMonthForDate,
  getFirstDateOfMonth,
  getLastDateOfMonth,
  getDifferenceFromMonday,
  getDifferenceFromSunday,
  isBefore,
  isBeforeOrEqual,
  lighten,
  darken
};
