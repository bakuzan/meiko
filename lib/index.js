// Components
export { default as Alert } from './Alert';
export { default as AppInformation } from './AppInformation';
export { default as AutocompleteInput } from './AutocompleteInput';
export { Button, withButtonisation } from './Button';
export { default as ChipListInput } from './ChipListInput';
export { default as ClearableInput } from './ClearableInput';
export { default as DateSelector } from './DateSelector';
export { default as Dialog } from './Dialog';
export { default as DropdownMenu } from './DropdownMenu';
export { default as FileUploader } from './FileUploader';
export { default as Form } from './Form';
export { default as FormControls } from './FormControls';
export { default as Grid } from './Grid';
export { default as Header } from './Header';
export { default as Image } from './Image';
export { default as ImageSelector } from './ImageSelector';
export { default as List } from './List';
export { SimpleLoading, Loading } from './Loadable';
export { default as LoadableContent } from './LoadableContent';
export { default as LoadingBouncer } from './LoadingBouncer';
export { default as LoadingSpinner } from './LoadingSpinner';
export { default as Logo } from './Logo';

export { default as MultiSelect } from './MultiSelect';
export { default as RadioButton } from './RadioButton';
export { default as RadioToggle } from './RadioToggle';
export { default as RatingControl } from './RatingControl';
export { default as RequestIndicator } from './RequestIndicator';
export { default as ScrollTopButton } from './ScrollTopButton';
export { default as SelectBox } from './SelectBox';
export { default as Sidebar } from './Sidebar';
export { default as Tabs } from './Tabs';
export { default as TagCloudSelector } from './TagCloudSelector';
export { default as Tickbox } from './Tickbox';
export { default as Toaster } from './Toaster';
export { default as Tooltip } from './Tooltip';

export { default as NewTabLink } from './NewTabLink';
export { default as Portal } from './Portal';

// Hooks
export { useGlobalStyles } from './hooks/useGlobalStyles';

// Utils
export { default as toaster, ToasterService } from './utils/toasterService';
export { default as addOutsideClick } from './utils/addOutsideClick';

export {
  PositionEnum,
  KeyCodes,
  OPEN_KEYS,
  CLOSE_KEYS,
  Icons,
  Strings,
  Types,
  Urls
} from './constants';
