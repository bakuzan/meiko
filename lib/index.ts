import './styles/index.scss';

import * as Components from './components';
import { default as fetchFromServer, handleErrorResponse } from './utils/fetch';
import * as Constants from './constants';

export default {
  ...Components,
  Constants,
  fetchFromServer,
  handleErrorResponse
};
