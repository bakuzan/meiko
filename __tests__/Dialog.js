import React from 'react';

import { Dialog } from '../lib';

const mockedActionFn = jest.fn();
const mockedCloseFn = jest.fn();

const firstId = 'test';
const lastId = 'test';
const tabTrapProps = { firstId, lastId };

it('should render with minimum props', function () {
  const { container } = render(
    <Dialog
      name="jest"
      isOpen={false}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(container.firstChild).toBeTruthy();
  expect(container).toMatchSnapshot();
});

it('should open when passed new prop', function () {
  const { container, rerender } = render(
    <Dialog
      name="jest"
      isOpen={false}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(container.firstChild.open).toBe(false);

  rerender(
    <Dialog
      name="jest"
      isOpen={true}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(container.firstChild.open).toBe(true);
  expect(container).toMatchSnapshot();
});

// todo must fix
xit('should call onCancel when cancel button clicked', async function () {
  const { container, getByText } = render(
    <Dialog
      name="jest"
      isOpen={true}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  const user = userEvent.setup();
  await user.click(getByText('Cancel'));

  expect(mockedCloseFn).toHaveBeenCalled();
  expect(container).toMatchSnapshot();
});

// todo must fix
xit('should render action button and call onAction when clicked', async function () {
  const { container, getByText } = render(
    <Dialog
      name="jest"
      isOpen={true}
      onAction={mockedActionFn}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  const user = userEvent.setup();
  await user.click(getByText('Submit'));

  expect(mockedActionFn).toHaveBeenCalled();
  expect(container).toMatchSnapshot();
});

it('should only render title when passed', function () {
  const { container, queryByRole, rerender } = render(
    <Dialog
      name="jest"
      isOpen={true}
      onAction={mockedActionFn}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(queryByRole('heading')).toBeNull();

  rerender(
    <Dialog
      name="jest"
      title={'jester!'}
      isOpen={true}
      onAction={mockedActionFn}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(queryByRole('heading')).toBeTruthy();
  expect(container).toMatchSnapshot();
});

it('should replace form element with div', function () {
  const { container, getByText, rerender } = render(
    <Dialog
      name="jest"
      isOpen={true}
      onAction={mockedActionFn}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(getByText('Submit').type).toEqual('submit');

  rerender(
    <Dialog
      name="jest"
      isOpen={true}
      isForm={false}
      onAction={mockedActionFn}
      onCancel={mockedCloseFn}
      tabTrapProps={tabTrapProps}
    >
      <div>jest</div>
    </Dialog>
  );

  expect(getByText('Submit').type).toEqual('button');
  expect(container).toMatchSnapshot();
});
