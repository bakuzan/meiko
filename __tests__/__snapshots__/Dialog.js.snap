// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call onCancel when cancel button clicked 1`] = `
<div>
  <dialog
    class="dialog dialog--backdrop  mko_1h7ugta-dialog  mko_1h7ugta-dialog_backdrop"
    open=""
    style="top: calc(0px + 50vh);"
  >
    <div>
      <input
        class="tab-trap tab-trap--top  mko_1q23t0y-tabTrap"
        type="text"
      />
      <div
        class="dialog__content"
      >
        <form
          autocomplete="off"
          class="dialog-content"
          name="jest"
          novalidate=""
        >
          <div
            class="dialog__content-custom"
          >
            <div>
              jest
            </div>
          </div>
          <div
            class="dialog__actions  mko_1h7ugta-actions"
          >
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__cancel"
              id="jestCancel"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
      <input
        class="tab-trap tab-trap--bottom  mko_1q23t0y-tabTrap"
        type="text"
      />
    </div>
  </dialog>
</div>
`;

exports[`should only render title when passed 1`] = `
<div>
  <dialog
    class="dialog dialog--backdrop  mko_1h7ugta-dialog  mko_1h7ugta-dialog_backdrop"
    open=""
    style="top: calc(0px + 50vh);"
  >
    <div>
      <input
        class="tab-trap tab-trap--top  mko_1q23t0y-tabTrap"
        type="text"
      />
      <div
        class="dialog__content"
      >
        <form
          autocomplete="off"
          class="dialog-content"
          name="jest"
          novalidate=""
        >
          <h4
            class="dialog__title  mko_1h7ugta-title"
          >
            jester!
          </h4>
          <div
            class="dialog__content-custom"
          >
            <div>
              jest
            </div>
          </div>
          <div
            class="dialog__actions  mko_1h7ugta-actions"
          >
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__action"
              id="jestAction"
              type="submit"
            >
              Submit
            </button>
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__cancel"
              id="jestCancel"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
      <input
        class="tab-trap tab-trap--bottom  mko_1q23t0y-tabTrap"
        type="text"
      />
    </div>
  </dialog>
</div>
`;

exports[`should open when passed new prop 1`] = `
<div>
  <dialog
    class="dialog dialog--backdrop  mko_1h7ugta-dialog  mko_1h7ugta-dialog_backdrop"
    open=""
    style="top: calc(0px + 50vh);"
  >
    <div>
      <input
        class="tab-trap tab-trap--top  mko_1q23t0y-tabTrap"
        type="text"
      />
      <div
        class="dialog__content"
      >
        <form
          autocomplete="off"
          class="dialog-content"
          name="jest"
          novalidate=""
        >
          <div
            class="dialog__content-custom"
          >
            <div>
              jest
            </div>
          </div>
          <div
            class="dialog__actions  mko_1h7ugta-actions"
          >
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__cancel"
              id="jestCancel"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
      <input
        class="tab-trap tab-trap--bottom  mko_1q23t0y-tabTrap"
        type="text"
      />
    </div>
  </dialog>
</div>
`;

exports[`should render action button and call onAction when clicked 1`] = `
<div>
  <dialog
    class="dialog dialog--backdrop  mko_1h7ugta-dialog  mko_1h7ugta-dialog_backdrop"
    open=""
    style="top: calc(0px + 50vh);"
  >
    <div>
      <input
        class="tab-trap tab-trap--top  mko_1q23t0y-tabTrap"
        type="text"
      />
      <div
        class="dialog__content"
      >
        <form
          autocomplete="off"
          class="dialog-content"
          name="jest"
          novalidate=""
        >
          <div
            class="dialog__content-custom"
          >
            <div>
              jest
            </div>
          </div>
          <div
            class="dialog__actions  mko_1h7ugta-actions"
          >
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__action"
              id="jestAction"
              type="submit"
            >
              Submit
            </button>
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__cancel"
              id="jestCancel"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
      <input
        class="tab-trap tab-trap--bottom  mko_1q23t0y-tabTrap"
        type="text"
      />
    </div>
  </dialog>
</div>
`;

exports[`should render with minimum props 1`] = `
<div>
  <dialog
    class="dialog dialog--backdrop  mko_1h7ugta-dialog  mko_1h7ugta-dialog_backdrop"
    style="top: calc(0px + 50vh);"
  >
    <div>
      <div
        class="dialog__content"
      >
        <form
          autocomplete="off"
          class="dialog-content"
          name="jest"
          novalidate=""
        >
          <div
            class="dialog__content-custom"
          >
            <div>
              jest
            </div>
          </div>
          <div
            class="dialog__actions  mko_1h7ugta-actions"
          >
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__cancel"
              id="jestCancel"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  </dialog>
</div>
`;

exports[`should replace form element with div 1`] = `
<div>
  <dialog
    class="dialog dialog--backdrop  mko_1h7ugta-dialog  mko_1h7ugta-dialog_backdrop"
    open=""
    style="top: calc(0px + 50vh);"
  >
    <div>
      <input
        class="tab-trap tab-trap--top  mko_1q23t0y-tabTrap"
        type="text"
      />
      <div
        class="dialog__content"
      >
        <div
          class="dialog-content"
        >
          <div
            class="dialog__content-custom"
          >
            <div>
              jest
            </div>
          </div>
          <div
            class="dialog__actions  mko_1h7ugta-actions"
          >
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__action"
              id="jestAction"
              type="button"
            >
              Submit
            </button>
            <button
              class="button ripple button--standard  mko_c5ujei-button  mko_c5ujei-button_standard dialog__cancel"
              id="jestCancel"
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
      <input
        class="tab-trap tab-trap--bottom  mko_1q23t0y-tabTrap"
        type="text"
      />
    </div>
  </dialog>
</div>
`;
